File: Sources/Services/JSONLParser.swift
import Foundation

/// Parses Claude Code JSONL session log files into TokenEvent records.
/// Location: ~/.claude/projects/<project>/<session>.jsonl
/// Each line is a JSON event. Corrupted lines are skipped gracefully.
struct JSONLParser {

    /// Parse a single JSONL line into a TokenEvent, if it contains token usage data.
    static func parseLine(_ line: String, project: String? = nil, sessionId: String? = nil) -> TokenEvent? {
        guard !line.isEmpty else { return nil }
        guard let data = line.data(using: .utf8) else { return nil }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        // Look for token usage in the message structure
        // JSONL events can have different shapes — we extract what we can
        let usage = extractUsage(from: json)
        guard usage.hasTokens else { return nil }

        let timestamp = extractTimestamp(from: json)
        let model = extractModel(from: json)
        let sid = extractSessionId(from: json) ?? sessionId
        let cost = extractCost(from: json)

        return TokenEvent(
            timestamp: timestamp,
            source: "claude_code",
            sessionId: sid,
            project: project,
            model: model,
            inputTokens: usage.input,
            outputTokens: usage.output,
            cacheCreate: usage.cacheCreate,
            cacheRead: usage.cacheRead,
            costUsd: cost
        )
    }

    /// Parse an entire JSONL file, returning all extractable token events.
    static func parseFile(at path: String, project: String? = nil) -> [TokenEvent] {
        guard let content = try? String(contentsOfFile: path, encoding: .utf8) else { return [] }

        let sessionId = sessionIdFromPath(path)
        let projectName = project ?? projectNameFromPath(path)

        return content.split(separator: "\n", omittingEmptySubsequences: true).compactMap { line in
            parseLine(String(line), project: projectName, sessionId: sessionId)
        }
    }

    /// Parse only new lines from a file, starting at the given byte offset.
    /// Returns all parsed events (with completeness flag) and the new offset.
    static func parseNewLines(at path: String, fromOffset offset: UInt64, project: String? = nil) -> (events: [(event: TokenEvent, isComplete: Bool)], newOffset: UInt64) {
        guard let handle = FileHandle(forReadingAtPath: path) else {
            return ([], offset)
        }
        defer { try? handle.close() }

        handle.seek(toFileOffset: offset)
        let data = handle.readDataToEndOfFile()
        let newOffset = offset + UInt64(data.count)

        guard let content = String(data: data, encoding: .utf8) else {
            return ([], newOffset)
        }

        let sessionId = sessionIdFromPath(path)
        let projectName = project ?? projectNameFromPath(path)

        let events: [(event: TokenEvent, isComplete: Bool)] = content.split(separator: "\n", omittingEmptySubsequences: true).compactMap { line in
            let lineStr = String(line)
            guard let event = parseLine(lineStr, project: projectName, sessionId: sessionId) else { return nil }
            return (event: event, isComplete: isCompleteEvent(lineStr))
        }

        return (events, newOffset)
    }

    // MARK: - Extraction helpers

    private struct Usage {
        var input: Int = 0
        var output: Int = 0
        var cacheCreate: Int = 0
        var cacheRead: Int = 0
        var hasTokens: Bool { input > 0 || output > 0 || cacheCreate > 0 || cacheRead > 0 }
    }

    private static func extractUsage(from json: [String: Any]) -> Usage {
        var usage = Usage()

        // Direct fields
        if let input = json["input_tokens"] as? Int { usage.input = input }
        if let output = json["output_tokens"] as? Int { usage.output = output }
        if let cc = json["cache_creation_input_tokens"] as? Int { usage.cacheCreate = cc }
        if let cr = json["cache_read_input_tokens"] as? Int { usage.cacheRead = cr }
        if usage.hasTokens { return usage }

        // Nested under "usage"
        if let u = json["usage"] as? [String: Any] {
            usage.input = u["input_tokens"] as? Int ?? 0
            usage.output = u["output_tokens"] as? Int ?? 0
            usage.cacheCreate = u["cache_creation_input_tokens"] as? Int ?? 0
            usage.cacheRead = u["cache_read_input_tokens"] as? Int ?? 0
            if usage.hasTokens { return usage }
        }

        // Nested under "message" → "usage"
        if let msg = json["message"] as? [String: Any],
           let u = msg["usage"] as? [String: Any] {
            usage.input = u["input_tokens"] as? Int ?? 0
            usage.output = u["output_tokens"] as? Int ?? 0
            usage.cacheCreate = u["cache_creation_input_tokens"] as? Int ?? 0
            usage.cacheRead = u["cache_read_input_tokens"] as? Int ?? 0
        }

        // Context window format (from Status hook)
        if let cw = json["context_window"] as? [String: Any],
           let cu = cw["current_usage"] as? [String: Any] {
            usage.input = cu["input_tokens"] as? Int ?? 0
            usage.output = cu["output_tokens"] as? Int ?? 0
            usage.cacheCreate = cu["cache_creation_input_tokens"] as? Int ?? 0
            usage.cacheRead = cu["cache_read_input_tokens"] as? Int ?? 0
        }

        return usage
    }

    private static func extractTimestamp(from json: [String: Any]) -> String {
        if let ts = json["timestamp"] as? String { return ts }
        if let ts = json["ts"] as? String { return ts }
        // Fallback: use current time
        let formatter = ISO8601DateFormatter()
        return formatter.string(from: Date())
    }

    private static func extractModel(from json: [String: Any]) -> String {
        if let model = json["model"] as? String { return model }
        if let model = json["model"] as? [String: Any] {
            return model["id"] as? String ?? model["api_model_id"] as? String ?? "unknown"
        }
        if let msg = json["message"] as? [String: Any],
           let model = msg["model"] as? String { return model }
        return "unknown"
    }

    private static func extractSessionId(from json: [String: Any]) -> String? {
        json["session_id"] as? String ?? json["sid"] as? String
    }

    private static func extractCost(from json: [String: Any]) -> Double? {
        if let cost = json["costUSD"] as? Double { return cost }
        if let cost = json["cost"] as? Double { return cost }
        if let costObj = json["cost"] as? [String: Any],
           let total = costObj["total_cost_usd"] as? Double { return total }
        return nil
    }

    // MARK: - Event Completeness

    /// Check if a JSONL line represents a completed message (has stop_reason).
    /// Intermediate streaming events have stop_reason=null and should not be
    /// stored in the DB to avoid double-counting, but can be used for real-time display.
    static func isCompleteEvent(_ line: String) -> Bool {
        guard let data = line.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return true // Non-message events (e.g. user, progress) are "complete"
        }
        // If it has a message with stop_reason=null, it's an intermediate streaming event
        if let msg = json["message"] as? [String: Any] {
            let stopReason = msg["stop_reason"]
            if stopReason == nil || stopReason is NSNull {
                return false
            }
        }
        return true
    }

    // MARK: - Path helpers

    /// Extract session ID from JSONL filename (filename without extension).
    static func sessionIdFromPath(_ path: String) -> String {
        (path as NSString).lastPathComponent.replacingOccurrences(of: ".jsonl", with: "")
    }

    /// Extract project name from JSONL path.
    /// Path format: ~/.claude/projects/<encoded-project-path>/<session>.jsonl
    static func projectNameFromPath(_ path: String) -> String? {
        let components = (path as NSString).pathComponents
        // Find "projects" in path and take the next component
        if let idx = components.firstIndex(of: "projects"), idx + 1 < components.count {
            let encoded = components[idx + 1]
            // The directory name is the project path with slashes replaced by hyphens
            // Return just the last meaningful segment
            let parts = encoded.split(separator: "-")
            if let last = parts.last { return String(last) }
            return encoded
        }
        return nil
    }
}

File: Windows/src-tauri/src/jsonl_parser.rs
use crate::database::TokenEvent;
use serde_json::Value;
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};

/// Result of parsing a single JSONL line.
pub struct ParsedEvent {
    pub event: TokenEvent,
    /// True if stop_reason is set (completed event → store in DB).
    /// False for intermediate streaming events (display only).
    pub is_complete: bool,
}

/// Parse a single JSONL line into a TokenEvent, if it contains token usage data.
pub fn parse_line(line: &str, project: Option<&str>, session_id: Option<&str>) -> Option<ParsedEvent> {
    if line.trim().is_empty() {
        return None;
    }
    let json: Value = serde_json::from_str(line).ok()?;
    let usage = extract_usage(&json);
    if !usage.has_tokens() {
        return None;
    }

    let timestamp = extract_timestamp(&json);
    let model = extract_model(&json);
    let sid = extract_session_id(&json).or_else(|| session_id.map(String::from));
    let cost = extract_cost(&json);
    let is_complete = is_complete_event(&json);

    Some(ParsedEvent {
        event: TokenEvent {
            id: None,
            timestamp,
            source: "claude_code".to_string(),
            session_id: sid,
            project: project.map(String::from),
            model,
            input_tokens: usage.input,
            output_tokens: usage.output,
            cache_create: usage.cache_create,
            cache_read: usage.cache_read,
            cost_usd: cost,
        },
        is_complete,
    })
}

/// Parse new lines from a file starting at the given byte offset.
/// Returns parsed events and the new byte offset.
pub fn parse_new_lines(
    path: &str,
    offset: u64,
    project: Option<&str>,
) -> (Vec<ParsedEvent>, u64) {
    let mut file = match File::open(path) {
        Ok(f) => f,
        Err(_) => return (Vec::new(), offset),
    };

    if file.seek(SeekFrom::Start(offset)).is_err() {
        return (Vec::new(), offset);
    }

    let mut buf = String::new();
    let bytes_read = match file.read_to_string(&mut buf) {
        Ok(n) => n as u64,
        Err(_) => return (Vec::new(), offset),
    };
    let new_offset = offset + bytes_read;

    let session_id = session_id_from_path(path);
    let proj_from_path = project_name_from_path(path);
    let proj = project.or(proj_from_path.as_deref()).map(String::from);

    let events: Vec<ParsedEvent> = buf
        .lines()
        .filter(|l| !l.trim().is_empty())
        .filter_map(|line| parse_line(line, proj.as_deref(), session_id.as_deref()))
        .collect();

    (events, new_offset)
}

// ── Usage extraction ──────────────────────────────────────────────

struct Usage {
    input: i64,
    output: i64,
    cache_create: i64,
    cache_read: i64,
}

impl Usage {
    fn has_tokens(&self) -> bool {
        self.input > 0 || self.output > 0 || self.cache_create > 0 || self.cache_read > 0
    }
}

fn extract_usage(json: &Value) -> Usage {
    // Direct fields
    let mut u = Usage {
        input: json["input_tokens"].as_i64().unwrap_or(0),
        output: json["output_tokens"].as_i64().unwrap_or(0),
        cache_create: json["cache_creation_input_tokens"].as_i64().unwrap_or(0),
        cache_read: json["cache_read_input_tokens"].as_i64().unwrap_or(0),
    };
    if u.has_tokens() {
        return u;
    }

    // Nested under "usage"
    if let Some(usage) = json.get("usage") {
        u = usage_from_obj(usage);
        if u.has_tokens() {
            return u;
        }
    }

    // Nested under "message" → "usage"
    if let Some(msg) = json.get("message") {
        if let Some(usage) = msg.get("usage") {
            u = usage_from_obj(usage);
            if u.has_tokens() {
                return u;
            }
        }
    }

    // Nested under "result" → "usage"
    if let Some(result) = json.get("result") {
        if let Some(usage) = result.get("usage") {
            u = usage_from_obj(usage);
            if u.has_tokens() {
                return u;
            }
        }
    }

    // Nested under "response" → "usage"
    if let Some(resp) = json.get("response") {
        if let Some(usage) = resp.get("usage") {
            u = usage_from_obj(usage);
            if u.has_tokens() {
                return u;
            }
        }
    }

    // Context window format (from Status hook)
    if let Some(cw) = json.get("context_window") {
        if let Some(cu) = cw.get("current_usage") {
            u = usage_from_obj(cu);
            if u.has_tokens() {
                return u;
            }
        }
    }

    u
}

fn usage_from_obj(v: &Value) -> Usage {
    Usage {
        input: v["input_tokens"].as_i64().unwrap_or(0),
        output: v["output_tokens"].as_i64().unwrap_or(0),
        cache_create: v["cache_creation_input_tokens"].as_i64().unwrap_or(0),
        cache_read: v["cache_read_input_tokens"].as_i64().unwrap_or(0),
    }
}

fn extract_timestamp(json: &Value) -> String {
    if let Some(ts) = json["timestamp"].as_str() {
        return ts.to_string();
    }
    if let Some(ts) = json["ts"].as_str() {
        return ts.to_string();
    }
    chrono::Utc::now().to_rfc3339()
}

fn extract_model(json: &Value) -> String {
    if let Some(m) = json["model"].as_str() {
        return m.to_string();
    }
    if let Some(obj) = json["model"].as_object() {
        if let Some(id) = obj.get("id").and_then(|v| v.as_str()) {
            return id.to_string();
        }
        if let Some(id) = obj.get("api_model_id").and_then(|v| v.as_str()) {
            return id.to_string();
        }
    }
    if let Some(msg) = json.get("message") {
        if let Some(m) = msg["model"].as_str() {
            return m.to_string();
        }
    }
    "unknown".to_string()
}

fn extract_session_id(json: &Value) -> Option<String> {
    json["session_id"]
        .as_str()
        .or_else(|| json["sid"].as_str())
        .map(String::from)
}

fn extract_cost(json: &Value) -> Option<f64> {
    if let Some(c) = json["costUSD"].as_f64() {
        return Some(c);
    }
    if let Some(c) = json["cost"].as_f64() {
        return Some(c);
    }
    if let Some(obj) = json["cost"].as_object() {
        if let Some(c) = obj.get("total_cost_usd").and_then(|v| v.as_f64()) {
            return Some(c);
        }
    }
    None
}

/// Check if a JSON event represents a completed message (has stop_reason set).
fn is_complete_event(json: &Value) -> bool {
    if let Some(msg) = json.get("message") {
        if let Some(msg_obj) = msg.as_object() {
            match msg_obj.get("stop_reason") {
                None => return false,
                Some(v) if v.is_null() => return false,
                _ => return true,
            }
        }
    }
    true // Non-message events are considered complete
}

// ── Path helpers ──────────────────────────────────────────────────

/// Extract session ID from JSONL filename.
fn session_id_from_path(path: &str) -> Option<String> {
    std::path::Path::new(path)
        .file_stem()
        .and_then(|s| s.to_str())
        .map(String::from)
}

/// Extract project name from JSONL path.
/// Path format: ~/.claude/projects/<encoded-project-path>/<session>.jsonl
fn project_name_from_path(path: &str) -> Option<String> {
    let components: Vec<&str> = path.split(['/', '\\']).collect();
    if let Some(idx) = components.iter().position(|&c| c == "projects") {
        if idx + 1 < components.len() {
            let encoded = components[idx + 1];
            // Return the last meaningful segment of the directory name
            if let Some(last) = encoded.rsplit('-').next() {
                if !last.is_empty() {
                    return Some(last.to_string());
                }
            }
            return Some(encoded.to_string());
        }
    }
    None
}
